sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,318 @@
+        .header("X-aws-ec2-metadata-token-ttl-seconds", "60")
+        .send()
+        .context(error::ImdsRequest { method: "PUT", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })?;
+
+    info!("Retrieving list of available public keys from IMDS");
+    let public_key_list =
+        get_text_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URL, &imds_session_token)?;
+    debug!("available public keys '{}'", &public_key_list);
+
+    info!("Generating endpoint urls to retrieve text of available public keys");
+    let public_key_endpoints = get_public_key_endpoints_from_imds(&public_key_list)?;
+
+    info!("Retrieving public keys from IMDS");
+    let mut public_keys = Vec::new();
+    for endpoint in public_key_endpoints.iter() {
+        let public_key_text = get_text_from_imds(&client, endpoint.as_str(), &imds_session_token)?;
+        // Simple check to see if the text is probably an ssh key.
+        if &public_key_text[0..3] == "ssh" {
+            debug!("{}", &public_key_text.trim_end());
+            public_keys.push(public_key_text.trim_end().to_string())
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid key. Skipping...",
+                &public_key_text.trim_end()
+            );
+            continue;
+        }
+    }
+    ensure!(!public_keys.is_empty(), error::NoValidImdsKeys);
+
+    Ok(public_keys)
+}
+
+/// Returns a list of endpoint urls for the public keys in IMDS. Since IMDS returns the list of
+/// available public keys as '0=my-public-key', we need to strip the index from the list and
+/// embed it into the endpoint url.
+fn get_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<Url>> {
+    let mut public_key_endpoints = Vec::new();
+    for available_key in public_key_list.lines() {
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<i32>().is_ok() {
+            let public_key_endpoint = &format!("{}{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
+            let public_key_endpoint_url =
+                Url::parse(&public_key_endpoint).context(error::HttpUrlParse {
+                    url: public_key_endpoint,
+                })?;
+            debug!("{}", &public_key_endpoint);
+            public_key_endpoints.push(public_key_endpoint_url);
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid index. Skipping...",
+                &f[0]
+            );
+            continue;
+        }
+    }
+    ensure!(
+        !public_key_endpoints.is_empty(),
+        error::NoValidImdsKeyEndpoints
+    );
+    Ok(public_key_endpoints)
+}
+
+/// Print usage message.
+fn usage() -> ! {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!("Usage: {} user-data", program_name);
```

Comment:
I think this is incorrect - we want the user to supply the setting  to fetch, so the invocation would look like:

`shibaken settings.host-containers.<container name>.user-data`

Take a peek at schnauzer's invocation in `sources/models/shared-defaults/defaults.toml`